src/lib/rng/jitter_rng/jitter_rng.h

```diff
@@ -0,0 +1,39 @@
+/*
+* CPU Jitter Random Number Generator
+* (C) 2024 Planck Security S.A.
+*
+* Botan is released under the Simplified BSD License (see license.txt)
+*/
+
+#ifndef BOTAN_JITTER_RNG_H_
+#define BOTAN_JITTER_RNG_H_
+
+#include <botan/rng.h>
+
+namespace Botan {
+
+/*
+* RNG using libjitterentropy.
+*/
+class BOTAN_PUBLIC_API(3, 6) Jitter_RNG final : public RandomNumberGenerator {
+   public:
+      Jitter_RNG();
+
+      std::string name() const override;
+
+      bool is_seeded() const override;
+
+      bool accepts_input() const override;
+
+      void clear() override;
+
+   protected:
+      void fill_bytes_with_input(std::span<uint8_t> out, std::span<const uint8_t> in) override;
+
+   private:
+      struct Rand_Data;
+      std::unique_ptr<Rand_Data, void (*)(Rand_Data*)> m_jitter;
```

Comment:
Since you weren't liking the raw `delete p`, I got a better solution for this. I'm confident to say this should be the way to go.
### In short
You need to declare the destrutor of `Jitter_RNG` in the header and implement it in the source file. Then you can use the smart pointer as usual (no weird deleter or creation) without errors:
```
// in header, inside Jitter_RNG definition
public:
   ~Jitter_RNG(); // add this line
   Jitter_RNG();
...
private:
   struct Rand_Data;
   std::unique_ptr<Rand_Data> m_jitter; // normal smart pointer type with default deleter type
```
```
// in source file
Jitter_RNG::~Jitter_RNG() = default; // add this line
Jitter_RNG::Jitter_RNG() : m_jitter{std::make_unique<Rand_Data>()} {} // usual smart pointer creation
```
### Explanation
Turns out, the actual reason for the error you were seeing before, about `Rand_Data` being incomplete, is because `~unique_ptr<Rand_Data>` calls `delete` for it's internal `Rand_Data *` and that call requires `sizeof(Rand_Data)` to be valid. Calls to that `~unique_ptr<Rand_Data>` are implicitly generated inside both the constructor and destructor of `Jitter_RNG` (in the constructor, the destructors of the members are called if an exception is thrown, to avoid leaving members initialized which would lead to memory leaks).
Since the `~Jitter_RNG` was not explicitly defined, it was *implicitly defaulted as an inline function*. This means it's definition is available to all translation units that use it (this condition allows inlining). And in turn, this means that calls to `~unique_ptr<Rand_Data>` are potentially generated in translation units that do not have access to the definition of `Rand_Data`.
By preventing the destructor (and the constructor) from being inline functions, you effectively encapsulate all calls to `~unique_ptr<Rand_Data>` inside the translation unit that defines those functions and `Rand_Data`.

Best regards.

Reply:
Yes, that works, thank you very much again. Updated.